example/bwmf/bwmf_task.go

```diff
@@ -31,168 +33,516 @@ Topology: the topology is different from task to task. Each task will consider i
+
+	bt.stopCriteria = taskgraph_op.MakeFixCountStopCriteria(int(bt.numOfIters))
+	proj_len := 0
+	if bt.m > bt.n {
+		proj_len = bt.m * bt.k
+	} else {
+		proj_len = bt.n * bt.k
 	}
-	// if we have all data, start next iteration.
-	ctx.IncEpoch()
-}
 
-// Other nodes has served with their local shards.
-func (t *bwmfTask) ChildDataReady(ctx taskgraph.Context, childID uint64, req string, resp []byte) {
-	t.dtReady.CountDown()
+	bt.optimizer = taskgraph_op.NewProjectedGradient(
+		taskgraph_op.NewProjection(
+			taskgraph_op.NewAllTheSameParameter(1e20, proj_len),
+			taskgraph_op.NewAllTheSameParameter(1e-9, proj_len),
+		),
+		bt.beta,
+		bt.sigma,
+		bt.alpha,
+	)
+
+	// channels for event-based `run()`
+	bt.epochChange = make(chan *event, 1)
+	bt.getD = make(chan *event, 1)
+	bt.getT = make(chan *event, 1)
+	bt.dataReady = make(chan *event, 1)
+	bt.updateDone = make(chan *event, 1)
+	bt.exitChan = make(chan struct{})
+
+	bt.logger.Println("RowShard:", *bt.rowShard)
+	bt.logger.Println("ColumnShard:", *bt.columnShard)
+	bt.logger.Println("m: ", bt.m)
+	bt.logger.Println("n: ", bt.n)
+	bt.logger.Println("k: ", bt.k)
+	bt.logger.Println("blockId: ", bt.blockId)
+	bt.logger.Println("sharded T: ", bt.shardedT)
+	bt.logger.Println("sharded D: ", bt.shardedD)
+	bt.logWriter.Flush()
+
+	// Now we have everything initialized.
+	go bt.run()
 }
 
-// get request of D/T shards from others. Serve with local shard.
-func (t *bwmfTask) ServeAsParent(fromID uint64, req string, dataReceiver chan<- []byte) {
-	var b []byte
-	var err error
+type event struct {
+	ctx      context.Context
+	epoch    uint64
+	fromID   uint64
+	request  *pb.Request
+	retT     chan *pb.DenseMatrixShard
+	retD     chan *pb.DenseMatrixShard
+	response proto.Message
+	// to be extended
+}
 
-	go func() {
-		if t.epoch%2 == 0 {
-			b, err = json.Marshal(t.dShard)
+func (bt *bwmfTask) run() {
+	for {
+		select {
+		case ec := <-bt.epochChange:
+			bt.logger.Println("SELECT- Epoch change to ", ec.epoch, " at task ", bt.taskID)
+			bt.doEnterEpoch(ec.ctx, ec.epoch)
+
+		case reqT := <-bt.getT:
+			bt.logger.Println("SELECT- GetT at epoch", bt.epoch)
+			// XXX: tShard is guaranteed to have been updated here.
+			err := bt.framework.CheckGRPCContext(reqT.ctx)
 			if err != nil {
-				t.logger.Fatalf("Slave can't encode dShard error: %v\n", err)
+				bt.logger.Panicf("GetTShard grpc broken in context at task %d for epoch %d.", bt.taskID, bt.epoch)
+				close(reqT.retT)
 			}
-		} else {
-			b, err = json.Marshal(t.tShard)
+			bt.logger.Printf("To respond with retT at task %d epoch %d.", bt.taskID, bt.epoch)
+			reqT.retT <- bt.shardedT
+			bt.logger.Printf("Responded with retT at task %d epoch %d.", bt.taskID, bt.epoch)
+		case reqD := <-bt.getD:
+			bt.logger.Println("SELECT- GetD at epoch", bt.epoch)
+			// XXX: tShard is guaranteed to have been updated here.
+
+			bt.logger.Println("Before checking GRPC status...")
+			bt.logWriter.Flush()
+			err := bt.framework.CheckGRPCContext(reqD.ctx)
```

Comment:
It hangs there forever when the req is handle with a relative bigger latency. @fengjingchao